src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt

```diff
@@ -91,24 +93,27 @@ class ImplLookup(
             return _ctx!!
         }
 
-    fun findImplsAndTraits(ty: Ty): Collection<RsTraitOrImpl> {
+    fun findImplsAndTraits(ty: Ty): Set<RsTraitOrImpl> {
         return findImplsAndTraitsCache.getOrPut(project, freshen(ty)) { rawFindImplsAndTraits(ty) }
     }
 
-    private fun rawFindImplsAndTraits(ty: Ty): Collection<RsTraitOrImpl> {
-        return when (ty) {
-            is TyTypeParameter -> ty.getTraitBoundsTransitively().map { it.element }
-            is TyTraitObject -> BoundElement(ty.trait).flattenHierarchy.map { it.element }
+    private fun rawFindImplsAndTraits(ty: Ty): Set<RsTraitOrImpl> {
```

Comment:
As I see, nothing changed here except we now use `Set` instead of `List`. What are the reasons for this? I'm worried about this also because the returned collection always stores to the cache, so it is long-leaving object, and HashSet consumes much more memoty than plain ArrayList.

Reply:
Semantically, we want a `Set` of impls here. Is there a compact array-based readonly set implementation somewhere?

Reply:
Furthermore, `mutableSetOf` used there, so actually it's the `LinkedHashSet` =)

@matklad We don't have something like `TObjectHashSet` (but have in `Koloboke` collections! which we don't use...). Anyway I think that semantic meaning is secondary in as hot place as this.

Reply:
Hm, but can we have both proper semantic & good performance? 

```
class ArraySet<T>(private val data: Array<T>) : Set<T> {
    override val size: Int get() = data.size
    override fun contains(element: T): Boolean = data.contains(element)
    override fun containsAll(elements: Collection<T>): Boolean = elements.all { data.contains(it) }
    override fun isEmpty(): Boolean = data.isEmpty()
    override fun iterator(): Iterator<T> = data.iterator()

    companion object {
        inline fun <reified T> fromSet(s: Set<T>): ArraySet<T> = ArraySet(s.toTypedArray())
    }
}
```

> so actually it's the LinkedHashSet =)

Being deterministic is a nice property, for tests especially. So, would constructing a LinkedHashSet and then converting it to an ArraySet be OK performance-wise?

Reply:
@matklad As I know, java's `Set` guarantees O(1) for `contains`... And it looks rather like a hack than more beautiful solution =)

Reply:
>As I know, java's Set must provide O(1) for contains.

Even `TreeSet`? :trollface:

Reply:
Hmm, but not O(n) anyway! =D

Reply:
About @Undin's description for `Set` in the comment below:
Hmm, it's reasonable. It would be cool to add some tests for method lookup with `fn foo<T: Trait>(t: T) where T: Trait { ... }`. Without a `Set` we will run into ambiguity?